tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc

```diff
@@ -120,6 +120,139 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
   return true;
 }
 
+namespace {
+bool Is1dSliceWithoutStrides(const HloInstruction* instr) {
+  return instr->opcode() == HloOpcode::kSlice &&
+         1 == instr->slice_starts().size() &&
+         1 == instr->slice_limits().size() &&
+         1 == instr->slice_strides().size() &&
+         1 == instr->slice_strides().at(0);
+}
+
+bool IsSliceInputFusion(const HloInstruction& unnested_hlo) {
+  if (!unnested_hlo.IsInputFusion()) {
+    return false;
+  }
+
+  const HloInstruction* root = unnested_hlo.fused_expression_root();
+  if (Is1dSliceWithoutStrides(root)) {
+    return true;
+  }
+
+  if (root->opcode() != HloOpcode::kTuple) {
+    return false;
+  }
+
+  return absl::c_all_of(root->operands(), [](const HloInstruction* instr) {
+    return Is1dSliceWithoutStrides(instr);
+  });
+}
+
+bool ConcatHasNoEffect(const HloInstruction* concat) {
```

Comment:
Question: what does the codegen look like in this case? Is it also able to figure out it's a noop, fuse it all, and don't emit anything extra?

Reply:
In the case of HorizontalLoopFusion, the concat is inserted intentionally to increase the parallelism degree (by combining several fusions). So the concat will still be generated in the codegen.

The concat is simply generated by the elemental emitter.